Mathlib/Geometry/Manifold/Instances/Icc.lean

```diff
@@ -0,0 +1,201 @@
+lemma contMDiff_subtypeVal_Icc  :
+    ContMDiff (𝓡∂ 1) 𝓘(ℝ) n (fun (z : Icc x y) ↦ (z : ℝ)) := by
+  intro z
+  rw [contMDiffAt_iff]
+  refine ⟨by fun_prop, ?_⟩
+  simp? says
+    simp only [extChartAt, PartialHomeomorph.extend, PartialHomeomorph.refl_partialEquiv,
+      PartialEquiv.refl_source, PartialHomeomorph.singletonChartedSpace_chartAt_eq,
+      modelWithCornersSelf_partialEquiv, PartialEquiv.trans_refl, PartialEquiv.refl_coe,
+      Icc_chartedSpaceChartAt, PartialEquiv.coe_trans_symm, PartialHomeomorph.coe_coe_symm,
+      ModelWithCorners.toPartialEquiv_coe_symm, CompTriple.comp_eq, PartialEquiv.coe_trans,
+      ModelWithCorners.toPartialEquiv_coe, PartialHomeomorph.toFun_eq_coe, Function.comp_apply]
+  split_ifs with hz
+  · simp? [IccLeftChart, Function.comp_def, modelWithCornersEuclideanHalfSpace] says
+      simp only [IccLeftChart, Fin.isValue, PartialHomeomorph.mk_coe_symm, PartialEquiv.coe_symm_mk,
+      modelWithCornersEuclideanHalfSpace, ModelWithCorners.mk_symm, Function.comp_def,
+      Function.update_self, ModelWithCorners.mk_coe, PartialHomeomorph.mk_coe]
+    rw [Subtype.range_val_subtype]
+    have : ContDiff ℝ n (fun (z : EuclideanSpace ℝ (Fin 1)) ↦ z 0 + x) := by fun_prop
+    apply this.contDiffWithinAt.congr_of_eventuallyEq_of_mem; swap
+    · simpa using z.2.1
+    have : {w : EuclideanSpace ℝ (Fin 1) | w 0 < y - x} ∈ 𝓝 (fun i ↦ z - x) := by
+      apply (isOpen_lt (continuous_apply 0) continuous_const).mem_nhds
+      simpa using hz
+    filter_upwards [self_mem_nhdsWithin, nhdsWithin_le_nhds this] with w hw h'w
+    rw [max_eq_left hw, min_eq_left]
+    linarith
+  · simp only [not_lt] at hz
+    simp? [IccRightChart, Function.comp_def, modelWithCornersEuclideanHalfSpace] says
+      simp only [IccRightChart, Fin.isValue, PartialHomeomorph.mk_coe_symm,
+        PartialEquiv.coe_symm_mk, modelWithCornersEuclideanHalfSpace, ModelWithCorners.mk_symm,
+        Function.comp_def, Function.update_self, ModelWithCorners.mk_coe, PartialHomeomorph.mk_coe]
+    rw [Subtype.range_val_subtype]
+    have : ContDiff ℝ n (fun (z : EuclideanSpace ℝ (Fin 1)) ↦ y - z 0) := by fun_prop
+    apply this.contDiffWithinAt.congr_of_eventuallyEq_of_mem; swap
+    · simpa using z.2.2
+    have : {w : EuclideanSpace ℝ (Fin 1) | w 0 < y - x} ∈ 𝓝 (fun i ↦ y - z) := by
+      apply (isOpen_lt (continuous_apply 0) continuous_const).mem_nhds
+      simpa using h.out.trans_le hz
+    filter_upwards [self_mem_nhdsWithin, nhdsWithin_le_nhds this] with w hw h'w
+    rw [max_eq_left hw, max_eq_left]
+    linarith
+
+/-- The projection from `ℝ` to a closed segment is smooth on the segment, in the manifold sense. -/
+lemma contMDiffOn_projIcc :
+    ContMDiffOn 𝓘(ℝ) (𝓡∂ 1) n (Set.projIcc x y h.out.le) (Icc x y) := by
+  intro z hz
+  rw [contMDiffWithinAt_iff]
+  refine ⟨by apply ContinuousAt.continuousWithinAt; fun_prop, ?_⟩
+  simp? says
+    simp only [extChartAt, PartialHomeomorph.extend, Icc_chartedSpaceChartAt,
```

Comment:
Similar comments here: can this be compressed using mfld_simps; is a higher-level comment useful?